Fix null parsing on S3 notification email

albertisfu opened the pull request:
When parsing an S3 notification email as part of pre-processing some special characters like "=20" were removed and caused an error during parsing in some emails with a document URL like the following: 
https://ecf.cod.uscourts.gov/doc1/20914522284?caseid=206379&de_seq_num=204&magic_num=20105679  

Parameters starting with `20` like` caseid=206379`, `de_seq_num=204`,` magic_num=20105679` failed during parsing since `=20` were replaced by a white space

In order to solve this problem, before replacing `=20`, we replace temporarily url variables `caseid=` `de_seq_num=` `magic_num=` by  `caseid%` `de_seq_num%` `magic_num%`  then replace `=20` characters and finally restore url variables to `caseid=` `de_seq_num=` `magic_num=`

A special test case like the one described was added in order to verify the problem is gone.

johnhawkinson commented:
There's more than one problem with malformed HTML in CMECF NEFs. For instance, gmail renders these (see the email list in the screenshot) as literal `&` `n` `b` `s` `p` strings, presumably because of the lack of a terminating `;`. The point being, I suspect the problem is more general and worry this solution, as described, may be too narrow.

<img width="739" alt="Screen Shot 2022-02-24 at 11 27 07" src="https://user-images.githubusercontent.com/1270718/155565989-7efcd174-f32e-4121-b7bc-db7e89561e96.png">

mlissner commented:
@johnhawkinson if the emails contain `&nbsp` without a semi-colon, I think we do like gmail does and just ignore it. That seems fine. Is there another example that we'd want to deal with? 

@albertisfu, I'm a bit confused here why we have to do do the find/replace of all these funny characters in general. Aren't they only in the plaintext versions? For example, the test case you added, there are no `=20`'s. Maybe we only do this find and replace in plaintext emails or something like that?

albertisfu commented:
@mlissner yes I'm also confused about s3 email examples some of them the content is in plain text and others are in HTML, I think those ones in plain text that have some special characters need to be converted to: `=2E`->`.`  ,  `=C2=A0`->` `  ,  `=3D`->`=`   and  `=20` ->`  `

I looked for '=20' in s3 example emails and the only test case where a could find a `=20` character in content is this one https://github.com/freelawproject/juriscraper/blob/main/tests/examples/pacer/nef/s3/arwd_5.txt#L199 

`FileNumber=2082324-0`

This one is an HTML case so I don't know if in this case, it is correct to replace `=20` to be a white space because it would change `FileNumber 82324-0` but I think we are not getting this number when parsing.

So I'm not sure if it would exist a case where is necessary to  transform`=20` to a white space, if not maybe we could avoid replacing `=20`

In the test case I added, I used a URL like the one we had problems parsing because it contains `=20` : 
https://ecf.cod.uscourts.gov/doc1/20914522284?caseid**=20**6379&de_seq_num**=20**4&magic_num**=20**105679

So we this test we just verify that we avoid replacing those `=20` and parse correctly.

mlissner commented:
I was able to talk to @tewen about this a bit. The idea for the scraper at present is that it can work on both plaintext or HTML emails (it doesn't look like PACER has figured out what multi-part ones are yet!). 

I think that means we should:

1. Detect early on if an email is HTML or plaintext.
2. If plaintext, do the find/replace thing.
3. If not, don't.

It also occurs to me that we'll still have this bug in plaintext emails if we continue blindly replacing `=20` characters, so on top of only applying the find/replace to plaintext emails, we also need to do so more carefully. Maybe it only applies to `=20\n`, for example. 

Trevor is going to send some plaintext email examples tonight, but in the meantime, maybe we can just fix it so HTML works properly?

johnhawkinson commented:
> @johnhawkinson if the emails contain &nbsp without a semi-colon, I think we do like gmail does and just ignore it. That seems fine. Is there another example that we'd want to deal with?

Not offhand. It's just my recollection that CMECF HTML may be noncompliant.

But it sounds like a rather different problem here. Why would you be worrying about `=20` in plaintext emails? It seems like your MIME parser is wrong (or you're not using one?) Unless you're saying they are `Content-Transfer-Encoding: quoted-printable` and that is malformed?

For sure you should not be applying corrections that apply to one encoding to the other.

> (it doesn't look like PACER has figured out what multi-part ones are yet!)

I think CMECF (what's this PACER thing?) made the right design choice here, to give users a knob whether they get a plain-text email or an HTML email, rather than sending multipart. Especially when sometimes the emails are huge. Maybe we are at the point where multipart/alternative emails work **so well** that nothing else should ever be sent, but my experience is that email display and formatting has only been getting jankier with time, not better.

mlissner commented:
> or you're not using [a mime parser]

Yeah, I think that's probably the core of it. Alberto, do you think you could take a stab at upgrading this to use an email parser at the first step and then handle the HTML or the plaintext as a second step? 

This looks like a decent starting place: https://stackoverflow.com/questions/17874360/python-how-to-parse-the-body-from-a-raw-email-given-that-raw-email-does-not

What's your take?

albertisfu commented:
Sure I think that would be the right approach, let me check this, if I have any questions I'll be asking you.

tewen commented:
@albertisfu @mlissner and @johnhawkinson put 10 emails with both plain text and HTML up here: https://drive.google.com/file/d/1MiTgdukDlVHQwhcft2sMC1AZIXNZexr_/view?usp=sharing

albertisfu commented:
I'm aware that the goal of email parser is that it could work properly for HTML or plain text emails.

Reviewing current test cases and the last ones that Trevor sent us, I can see that all of those emails have a version in `text/html` and some of them also have a `text/plain` version in the same file, if I'm not wrong in those cases the `text/plain` version is added when the email is forwarded.

My questions are: 
- when getting emails from PACER, emails could have both versions (`text/html`and `text/plain`) in the same file or it's just possible to get one or another? 

- In case we get emails with both versions, which one should we prioritize to parse, `text/html` or `text/plain`?

johnhawkinson commented:
> Reviewing current test cases and the last ones that Trevor sent us, I can see that all of those emails have a version in `text/html` and some of them also have a `text/plain` version in the same file, if I'm not wrong in those cases the `text/plain` version is added when the email is forwarded.

Unfortunately, the method Trevor used to forward doesn't preserve encodings, so those are not very useful.

> when getting emails from PACER, emails could have both versions (`text/html`and `text/plain`) in the same file or it's just possible to get one or another?

CMECF sends emails either as `text/html` or as `text/plain` but not both.
Of course this could change in the future.

>  In case we get emails with both versions, which one should we prioritize to parse, `text/html` or `text/plain`?

I don't think it matters a whole lot, but CMECF does include markup in docket text, and that markup is lost in `text/plain`.
So all things being equal, prefer the html version. That said, Courtlistener does not do a good job of handling that markup (we strip it), which is bad, but it means it doesnt' matter much.

Digging through my mailbox, here's an old example of a `text/plain` email from back when I had NEFs configured in that fashion: [manafort2019.mbox.zip](https://github.com/freelawproject/juriscraper/files/8143677/manafort2019.mbox.zip)
.

albertisfu commented:
Understood, thanks @johnhawkinson

mlissner commented:
Thanks John! Yes, I agree with everything he said!

albertisfu commented:
MIME Parser email implemented, now it's not necessary to replace special characters.

Added support for parse text/plain emails and added a test case for it, I'm afraid that we are going to need more text/plain emails examples to ensure it works properly.

For example, I have a doubt about docket `description`, in this example, we have something like this:
```
Docket Text:
Joint  STATUS REPORT <i></i>   by USA
as to RICHARD W. GATES, III (Weissmann, Andrew)
```
Right now I'm parsing as: Joint  STATUS REPORT by USA
Is this correct?

johnhawkinson commented:
Huh! I had never noticed there was HTML markup in the `text/plain` NEFs. I wonder if that's still true, three years later, and I wonder if it's specific to `ecf.dcd`? I don't have a huge corpus of them. Although I could change some of my HTML notifications to `text/plain` and wait-and-see what the court drags in…

> Right now I'm parsing as: Joint STATUS REPORT by USA

Err, no? Did you inadvertently truncate? It should be

`Joint  STATUS REPORT by USA as to RICHARD W. GATES, III (Weissmann, Andrew)`

Just as it appears at https://www.courtlistener.com/docket/6249516/united-states-v-manafort/?filed_after=&filed_before=&entry_gte=551&entry_lte=551&order_by=asc

You could also, perhaps, draw the conclusion that `<i></i>` means there is an empty free text field here, but since we don't make use of that information, I'm not sure what to say other than keep it in the back of your head if you ever want it, just as you would when parsing the HTML version.

mlissner commented on juriscraper/pacer/email.py:
```diff
@@ -293,64 +403,48 @@ def _combine_lines_with_proper_spaces(self, text):
+                    self.content_type = "text/html"
+                    break
+        else:
+            # If not multipart, parse either text/html or text/html message
```
```suggestion
            # If not multipart, parse either text/html or text/plain message
```

mlissner commented on juriscraper/pacer/email.py:
```diff
@@ -293,64 +403,48 @@ def _combine_lines_with_proper_spaces(self, text):
+                    self.content_type = "text/plain"
+                    break
+
+        email_body = body.decode("UTF-8")
+        html_only = self._html_from_s3_email(email_body)
+
+        if self.content_type == "text/plain":
+            return super()._parse_text(email_body)
+        elif self.content_type == "text/html":
+            return super()._parse_text(html_only)
```
Why not like this:

```suggestion
        if self.content_type == "text/plain":
            email_body = body.decode("UTF-8")
            return super()._parse_text(email_body)
        elif self.content_type == "text/html":
            html_only = self._html_from_s3_email(email_body)
            return super()._parse_text(html_only)
```

albertisfu replied:
Solved, I just had to leave `email_body = body.decode("UTF-8")` outside because is used in both versions

mlissner reviewed (changes requested):
This looks pretty good to me, I think. I've sent a few comments already, and there's the one from John too, but I think that's it for now.

mlissner commented:
> I was checking the description parsing for text/html emails and I realized that it might not be working properly.

You're absolutely right. Thanks for catching that too! If you can get that fixed too, then we can merge?

albertisfu commented:
Sure, working on that fix.

albertisfu commented:
@mlissner Done! fixed `description` parsing in `text/html` emails, and updated JSON of test cases where the description was incomplete.

mlissner commented:
Looks great to me. Let's chat with Bill about getting this deployed once he's back from being sick.
